### ISSUE TITLE
Invariant failed: Side effects like changing state are not allowed at this point.

### ISSUE THREAD

**Issue by chobo2** [mobxjs/mobx-react#285]
Hi

I am getting this error

```
Uncaught Error: [mobx] Invariant failed: Side effects like changing state are not allowed at this point. Are you trying to modify state from, for example, the render function of a React component? Tried to modify: UserStore@2.users
    at invariant (app.bundle.js:8725)
    at fail (app.bundle.js:8720)
    at checkIfStateModificationsAreAllowed (app.bundle.js:9263)
    at ObservableArrayAdministration.spliceWithArray (app.bundle.js:6836)
    at ObservableArray.push (app.bundle.js:7020)
    at UserPage.render (app.bundle.js:35894)
    at Object.allowStateChanges (app.bundle.js:7439)
    at app.bundle.js:12736
    at trackDerivedFunction (app.bundle.js:9281)
    at Reaction.track (app.bundle.js:9449)
invariant @ app.bundle.js:8725
fail @ app.bundle.js:8720
checkIfStateModificationsAreAllowed @ app.bundle.js:9263
ObservableArrayAdministration.spliceWithArray @ app.bundle.js:6836
ObservableArray.push @ app.bundle.js:7020
render @ app.bundle.js:35894
allowStateChanges @ app.bundle.js:7439
(anonymous) @ app.bundle.js:12736
trackDerivedFunction @ app.bundle.js:9281
Reaction.track @ app.bundle.js:9449
reactiveRender @ app.bundle.js:12731
(anonymous) @ app.bundle.js:31984
measureLifeCyclePerf @ app.bundle.js:31264
_renderValidatedComponentWithoutOwnerOrContext @ app.bundle.js:31983
_renderValidatedComponent @ app.bundle.js:32010
_updateRenderedComponent @ app.bundle.js:31934
_performComponentUpdate @ app.bundle.js:31912
updateComponent @ app.bundle.js:31833
performUpdateIfNecessary @ app.bundle.js:31749
performUpdateIfNecessary @ app.bundle.js:4251
runBatchedUpdates @ app.bundle.js:1770
perform @ app.bundle.js:5748
perform @ app.bundle.js:5748
perform @ app.bundle.js:1709
flushBatchedUpdates @ app.bundle.js:1792
closeAll @ app.bundle.js:5814
perform @ app.bundle.js:5761
batchedUpdates @ app.bundle.js:33076
batchedUpdates @ app.bundle.js:1717
dispatchEvent @ app.bundle.js:33234
```

Maybe I don't have a solid understanding yet of how everything works and what I am doing is a bad idea but basically what I am trying to do is this.

User clicks a "div" that record is stored as "activeUser" and then I try to modify stuff in that object.

User Class

```
import {observable} from 'mobx';

class User {
    @observable firstName
    @observable lastName

    constructor(firstName = 'Jim', asselastNametType = '') {
        this.firstName = firstName;
        this.lastName = asselastNametType;
    }
}

export default User;
```
userStore

```
import {observable, computed, action} from 'mobx';
import User from '../classes/User'

class UserStore {
    @observable users = []
    @observable activeUser = new User();

    @computed get count() {
        return this.users.length;
    }
    @action setActiveUser(user){
        // I would like to pass in a paramter but not sure if it is possible with @action.bound
        user.firstName = 'test';
        this.activeUser = user;
      // also tried this but same error
       // this.activeUser.firstName = 'test';
        
    }
}

var userStore = new UserStore();

export default userStore;

```
UserPage

```
import React, { Component } from 'react'
import ReactDOM from 'react-dom';
import { observer, inject } from 'mobx-react';
import DevTools from 'mobx-react-devtools';

import User from '../classes/User'


@inject('userStore')
@observer
export default class UserPage extends Component {
    constructor() {
        super();
    }
    render() {
        let userStore = this.props.userStore;
        userStore.users.push(new User());
        return (
            <div style={{backgroundColor: "#FF0000"}} onClick={ () => userStore.setActiveUser(userStore.users[0])}>
                test {userStore.users[0].firstName}
                {userStore.activeUser.firstName}
            </div>
        )
    }
}

```

app.js

```
import React from 'react';
import ReactDOM from 'react-dom';
import stores from './stores';
import UserPage from './components/UserPage'
import { Provider  } from 'mobx-react';

ReactDOM.render(
    <Provider userStore={stores.userStore}>
        <UserPage  />
    </Provider>,      
       document.getElementById('root')
);
```

**Comment by andykog** [comment]
@chobo2, you are trying to modify state during `render()` with `userStore.users.push(new User());`, that's what error is about. Try using `componentWillMount`/`componentWillUpdate`.

**Comment by chobo2** [comment]
Thanks!

That makes sense but I would have thought that would have happened on initial render as well.

Also how about my other question regarding `@action.bound,` is it possible to pass in a parameter when using this?

**Comment by andykog** [comment]
> is it possible to pass in a parameter when using this?

Yes

**Comment by chobo2** [comment]
> is it possible to pass in a parameter when using this?
> Yes

Can you give me an example?

**Comment by andykog** [comment]
Here you go:
```js
@action.bound some(myParameter) {
  console.log(this, myParameter);
}
```

**Comment by chobo2** [comment]
> Here you go:
> 
> @action.bound some(myParameter) {
>   console.log(this, myParameter);
> }

How about the on click I did this `onClick={userStore.setActiveUser(userStore.users[0])}` but nothing happens now.

**Comment by andykog** [comment]
That's because you are calling a function and passing the result instad of function itself. Try `onClick={userStore.setActiveUser}` or `onClick={userStore.setActiveUser.bind(this, userStore.users[0])}`

**Comment by andykog** [comment]
Or `onClick={() => userStore.setActiveUser(userStore.users[0])}`

**Comment by chobo2** [comment]
@andykog  

Ok, thanks, so do you recommend if I got to pass something in should I just always then just use the arrow function even for ones that require nothing passed in?

Just not sure if it is a good idea to do sometimes the arrow function and sometimes `@action.bound.`

**Comment by tuancaraballo** [comment]
My problem was :  having two contexts in sibling components and one custom DragLayer per sibling, and where one draglayer  will superpose (as a modal/dialog) on top of an existing sibling dragLayer. For some reason, only one of the drag-layer would be "registered"

This was my solution:

1- Moved  @DragDropContext(HTML5Backend) to the parent component.
2- Set an observable array in my store (mobx) that kept track currently rendered DragLayers, then in the sibling components, I added a condition where only the component's DragLayer will render iff it was the last pushed DragLayer in the array. 
3- In every component, I added "push" component's draglayer or "pop" in the ComponentDidMount and ComponentWillUnmount life cycles, respectively. 

Hope it helps.
